ambry-store/src/main/java/com/github/ambry/store/IndexSegment.java

```diff
@@ -1149,10 +1158,11 @@ boolean getEntriesSince(StoreKey key, FindEntriesCondition findEntriesCondition,
    * @return an {@link Iterator<IndexEntry>}.
    */
   public Iterator<IndexEntry> iterator() {
-    if (!sealed.get()) {
-      return new UnsealedIndexSegmentEntryIterator();
+    NavigableMap<StoreKey, ConcurrentSkipListSet<IndexValue>> indexCopy = index;
+    if (!isSealedLocked()) {
+      return new UnsealedIndexSegmentEntryIterator(indexCopy);
     }
-    return new SealedIndexSegmentEntryIterator();
+    return new SealedIndexSegmentEntryIterator(serEntries);
```

Comment:
Can you be sure UnsealedIndexSegmentEntryIterator() is called when isSealed=false? What if another thread sets it to true as soon as isSealedLocked() returns. You need to get and hold the lock throughout this routine.